cpp/src/copying/copy.cu

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/wrappers/bool.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/detail/copy_if_else.cuh>
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+
+std::unique_ptr<column> copy_if_else( column_view const& boolean_mask, column_view const& lhs, column_view const& rhs,
+                                      rmm::mr::device_memory_resource *mr,
+                                      cudaStream_t stream)
+{
+   CUDF_EXPECTS(lhs.type() == rhs.type(), "Both columns must be of the same type");
+   CUDF_EXPECTS(lhs.size() == rhs.size(), "Both columns must be of the size");
+   CUDF_EXPECTS(boolean_mask.type() == data_type(BOOL8), "Boolean mask column must be of type BOOL8");   
+   CUDF_EXPECTS(boolean_mask.size() == lhs.size(), "Boolean mask column must be the same size as lhs and rhs columns");
+
+   // filter in this case is a column made of bools
+   auto bool_mask_device_ptr = column_device_view::create(boolean_mask);   
+   column_device_view bool_mask_device = *bool_mask_device_ptr;   
+   auto filter = [bool_mask_device] __device__ (cudf::size_type i) { return bool_mask_device.element<cudf::experimental::bool8>(i); }; 
+
+   return copy_if_else(filter, lhs, rhs, mr, stream);
```

Comment:
If `copy_if_else` is modified to take an iterator instead of a function object, then it can be simplified to just this. 
```suggestion
   return copy_if_else(boolean_mask.begin<experimental::bool8>(), lhs, rhs, mr, stream);
```

Reply:
Right, but Mark was saying he wanted a version that took a lambda.

Reply:
Iterator serves the same purpose of being more generic, but is also more idiomatic.

Reply:
True, but that is incongruous with copy_if. In fact it's incongruous with `std::copy_if`.

Reply:
Valid points. I had something like [`thrust::gather_if`](https://thrust.github.io/doc/group__gathering_gab487e68db682dd1b8437be2669a9b8c2.html#gab487e68db682dd1b8437be2669a9b8c2) in mind which accepts a separate `Stencil` iterator range that determines if each value should be gathered or not.